Adding a display_ytarray() function

MatthewTurk opened the pull request:
This is a draft, possibly-to-be-rejected, function that adds a `display_ytarray()` function.  At present, it is separately defined from the usual `YTArray` objects, as it only works with reasonably sized objects.  It's possible we could put it into the `YTArray` itself (and just bail on ipython display if the size is too big), but I'm not sure if we want to do that.

It looks like this by default:

![screenshot 2018-06-20 at 4 00 02 pm](https://user-images.githubusercontent.com/89019/41684429-0cfdaae4-74a3-11e8-9f8d-ddf2c14561b3.png)

The boxes are read-only.  Changing the displayed unit does *not* change the underlying array, but only the display.

![screenshot 2018-06-20 at 4 01 03 pm](https://user-images.githubusercontent.com/89019/41684484-26cf8e2e-74a3-11e8-978c-72841b13d512.png)

ngoldbaum commented:
Maybe it would make more sense to have a specific type we can use that has this defined on it as an `_ipython_display_` method? Something like `YTVectorArray` that specifically represents three-element vectors or something like that? I see why you might want to do something like this but I think that stuffing it into a helper function means that in practice no one will see that this functionality is even available, let alone use it.

ngoldbaum commented:
Also it's worth pointing out that I'm actively working on replacing yt.units with unyt in yt-4.0 so I think it makes sense not to try to put new stuff in `yt.units`.

MatthewTurk commented:
In an ideal world, this would live only on `YTIndexArray` objects -- but I don't know if that's going to happen.  I think having it on a vector of size 3 would not be any better than a helper function.  Sticking it in `ipython_display` and raising the error might be just as good, as it would cancel the ipython display on big arrays but use it elsewhere.

One reason I am not super concerned about visibility is that I want to use this internal to a dataset display, which would definitely implement the `_ipython_display_`.

MatthewTurk commented:
@ngoldbaum I know you are (and I know you know I know you are :grin: ), but the only way I know to make this accessible for the immediate future is via `master`.

Alternately, since I primarily want this as part of a larger widget, I could push all this into that widget.

ngoldbaum commented:
Ah I see, if this is an internal-only helper function then it's fine where it is.

My only other comment is that it's pretty wide by default, I'd consider collapsing this horizontally so it's a little easier to take in all at once in a single glance.

jzuhone commented:
I like this, but I'm also kind of interested in seeing exactly what your plans are for how you'd use it in more detail.

MatthewTurk commented:
@ngoldbaum I will try to do this, and to add a few other tests, without requiring a display venue for the widget.  One concern I have is that I don't want to redundantly test ipywidgets, but I think that's not so big a concern.

@jzuhone I've got another PR incoming that will specifically apply this to implement a `display(ds)` function, along with field browsing.

MatthewTurk commented:
I think the appveyor error is transitory and not related...?

ngoldbaum commented:
Looks like it. I just triggered another build.

cphyc commented on yt/units/yt_array.py:
```diff
@@ -1822,3 +1823,42 @@ def savetxt(fname, arrays, fmt='%.18e', delimiter='\t', header='',
+    r"""
+    Display a YTArray in a Jupyter widget that enables unit switching.
+
+    This 
```
There's something missing there

MatthewTurk replied:
Fixed!

MatthewTurk commented:
So I actually don't think I know how to fix this error -- it seems like there needs to be a Jupyter kernel for the tests.  Any ideas?

ngoldbaum commented:
according to https://github.com/ricklupton/ipysankeywidget/issues/8#issuecomment-267305451 this happens when trying to use traitlets with a widget outside of a jupyter kernel. sigh.... I wonder how e.g. ipywidgets tests this stuff.

MatthewTurk commented:
@yt-fido test this please

Xarthisius commented:
@yt-fido test this please

I was able to reproduce that in a clean docker env, when I ran tests with `ipywidgets==4.1.1`, which is the version we had installed on the build box. When I updated `ipywidgets` the issue went away. Let's see how this build goes.

ngoldbaum commented:
It might be nice to make `ds.domain_left_edge` and friends return objects that have a `_repr_html_` that calls this method. Doesn't need to be done now though.

You did give the function a docstring but it won't show up in the API docs unless you explicitly add an entry for it I think. It might also be worthwhile to mention this function somewhere in the narrative docs otherwise most people will probably never come across it.

In the future `YTArray` will be replaced by `unyt_array` so the name of this function might need to be changed then as well. It might be worth coming up with a name that doesn't explicitly use the word "ytarray". That said I'm not objecting to the name getting changed before merging.

MatthewTurk commented:
@ngoldbaum These are great comments.  To be honest, I'm *more* than happy to implement these -- this first pass was aiming at just trying to make it as unobtrusive as possible.  I'll implement:

 - [ ] `_repr_html_` attached to the DLE, DRE and DW attributes.
 - [ ] Making `display_ytarray` a method on `YTArray` itself, but call it `.display`.  This way we won't override `_repr_html_`'s default behavior, but it will show up.  (Unless you'd be okay with just implementing `_repr_html_` which I'd also be fine with; it's what pandas dataframes do.)

ngoldbaum commented:
One worry that I have about making it a method of `YTArray` is that it only works with arrays of up to 3 elements so in practice most arrays will not be able to be displayed. That might be confusing. That's why I suggested making the automatic display behavior for `domain_left_edge` and other attributes like it special, since we know that those will only have a certain number of elements.

MatthewTurk commented:
@ngoldbaum I followed your suggestions, but did not (yet) change the `display_ytarray` function name.  The reason I didn't was that -- as you point out -- it's mostly only useful right now for short arrays, *and* by design it's only used in interactive environments.  So I've added an `_ipython_display_` function that is added onto `MutableAttribute` instances where appropriate, and added `display_ytarray` to the API docs.

I think this might be ready to go?

Here's what it looks like for me.

![screenshot 2018-10-24 at 8 56 08 am](https://user-images.githubusercontent.com/89019/47436080-6c1b4880-d76b-11e8-9f6b-ece3634a1939.png)

MatthewTurk commented:
Am I reading the test results right and that the lint failure is in the parallel analysis interface?

Xarthisius commented:
@yt-fido test this please

Xarthisius commented:
@yt-fido test this please

MatthewTurk commented:
@yt-fido test this please

MatthewTurk commented:
@yt-fido test this please

brittonsmith commented:
I tried calling this function on an array of size greater than three and a `TypeError` exception was raised during the `YTArrayTooLargeToDisplay` exception. Here's the traceback:
http://paste.yt-project.org/show/116/
This seems to have fooled the test, but I don't know a way around that.

MatthewTurk commented:
@brittonsmith ah, good catch!  This is now fixed, thanks!

MatthewTurk commented:
This one (and #1848) are both candidates for rebasing on merge.
